### ISSUE TITLE
`create_before_destroy` resources display in plan as `destroy and then create`

### ISSUE THREAD

**Issue by M1ke** [hashicorp/terraform#17694]
I have already opened this bug at the provider level here:

https://github.com/terraform-providers/terraform-provider-aws/issues/3813

However it seems that the provider may not have the issue here, one comment on that bug says "The AWS provider doesn't control any of the resource ordering itself." so I am wondering if the problem is upstream from the provider.

### Terraform Version
v0.11.4 (also fails on 0.11.2)
Latest AWS provider (I ran `terraform init  -upgrade=true -get=true -get-plugins=true`)

### Affected Resource(s)
Please list the resources as a list, for example:
- aws_autoscaling_group
- aws_launch_configuration

### Terraform Configuration Files
```hcl
resource "aws_autoscaling_group" "dev" {
  desired_capacity = 1
  health_check_grace_period = 300
  health_check_type = "EC2"
  launch_configuration = "${aws_launch_configuration.dev.name}"
  max_size = 2
  min_size = 1
  name = "dev-${aws_launch_configuration.dev.name}"
  availability_zones = [
    "${data.aws_availability_zones.all.names[0]}",
    "${data.aws_availability_zones.all.names[1]}"]
  termination_policies = [
    "OldestLaunchConfiguration"]
  wait_for_capacity_timeout = "10m"
  metrics_granularity = "1Minute"
  target_group_arns = [
    "${aws_lb_target_group.TG.arn}"]
  min_elb_capacity = 1

  lifecycle {
    create_before_destroy = true
  }

  tag {
    key = "environment"
    propagate_at_launch = true
    value = "dev"
  }

  tag {
    key = "ami"
    propagate_at_launch = true
    value = "${var_ami_name}"
  }
}

resource "aws_launch_configuration" "dev" {
  image_id = "${var_ami}"
  instance_type = "t2.small"
  iam_instance_profile = "${aws_iam_role.ROLE.name}"
  key_name = "${var.KEY}"
  security_groups = [
    "${aws_security_group.SG.id}"]
  enable_monitoring = false
  ebs_optimized = false
  user_data = <<DATA
#!/bin/bash

# REDACTED FOR GITHUB ISSUE
DATA

  lifecycle {
    create_before_destroy = true
  }

  root_block_device {
    volume_type = "gp2"
    volume_size = 8
    delete_on_termination = true
  }
}
```

## Expected Behavior
Terraform chooses to create new resources before destroying. This has always been the behaviour previously.

### Actual Behavior
Terraform chooses to destroy first, then replace. No other changes have happened other than a new AMI (which is how I cycle in new AMIs to our ASG)

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform plan --target=aws_autoscaling_group.dev` (note this also happens for plan or apply in general)

### Important Factoids
Very basic set up, until today all apply/plans have worked to create/destroy in the expected order.

### Debug

Here's the output of `terraform plan --target=aws_autoscaling_group.dev`

To cause the change I added a single comment character to the user_data on the LC. Normally I'd cause the rebuild by changing the AMI ID when we want to make server config changes.

```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

aws_iam_role.dev-ec2: Refreshing state... (ID: dev-ec2)
aws_vpc.dev: Refreshing state... (ID: vpc-X)
aws_eip.dev-active: Refreshing state... (ID: eipalloc-Y)
data.aws_availability_zones.all: Refreshing state...
data.aws_ami.dev-ami: Refreshing state...
aws_security_group.dev-ec2: Refreshing state... (ID: sg-Z)
aws_lb_target_group.dev: Refreshing state... (ID: arn:aws:elasticloadbalancing:eu-west-1:...39528:targetgroup/dev/A)
aws_launch_configuration.dev: Refreshing state... (ID: terraform-20180316124712668100000001)
aws_autoscaling_group.dev: Refreshing state... (ID: dev-terraform-20180316124712668100000001)

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

-/+ aws_autoscaling_group.dev (new resource required)
      id:                                        "dev-terraform-20180316124712668100000001" => <computed> (forces new resource)
      arn:                                       "arn:aws:autoscaling:eu-west-1:X" => <computed>
      availability_zones.#:                      "2" => "2"
      availability_zones.1924028850:             "eu-west-1b" => "eu-west-1b"
      availability_zones.3953592328:             "eu-west-1a" => "eu-west-1a"
      default_cooldown:                          "300" => <computed>
      desired_capacity:                          "1" => "1"
      force_delete:                              "false" => "false"
      health_check_grace_period:                 "300" => "300"
      health_check_type:                         "EC2" => "EC2"
      launch_configuration:                      "terraform-20180316124712668100000001" => "${aws_launch_configuration.dev.name}"
      load_balancers.#:                          "0" => <computed>
      max_size:                                  "2" => "2"
      metrics_granularity:                       "1Minute" => "1Minute"
      min_elb_capacity:                          "1" => "1"
      min_size:                                  "1" => "1"
      name:                                      "dev-terraform-20180316124712668100000001" => "dev-${aws_launch_configuration.dev.name}" (forces new resource)
      protect_from_scale_in:                     "false" => "false"
      tag.#:                                     "2" => "2"
      tag.1447999130.key:                        "ami" => "ami"
      tag.1447999130.propagate_at_launch:        "true" => "true"
      tag.1447999130.value:                      "dev-DATE-1" => "dev-DATE-1"
      tag.3112388728.key:                        "environment" => "environment"
      tag.3112388728.propagate_at_launch:        "true" => "true"
      tag.3112388728.value:                      "dev" => "dev"
      target_group_arns.#:                       "1" => "1"
      target_group_arns.2885565262:              "arn:aws:elasticloadbalancing:Y" => "arn:aws:elasticloadbalancing:Y"
      termination_policies.#:                    "1" => "1"
      termination_policies.0:                    "OldestLaunchConfiguration" => "OldestLaunchConfiguration"
      vpc_zone_identifier.#:                     "0" => <computed>
      wait_for_capacity_timeout:                 "10m" => "10m"

-/+ aws_launch_configuration.dev (new resource required)
      id:                                        "terraform-20180316124712668100000001" => <computed> (forces new resource)
      associate_public_ip_address:               "false" => "false"
      ebs_block_device.#:                        "0" => <computed>
      ebs_optimized:                             "false" => "false"
      enable_monitoring:                         "false" => "false"
      iam_instance_profile:                      "dev-ROLE" => "dev-ROLE"
      image_id:                                  "ami-X" => "ami-X"
      instance_type:                             "t2.small" => "t2.small"
      key_name:                                  "KEY" => "KEY"
      name:                                      "terraform-20180316124712668100000001" => <computed>
      root_block_device.#:                       "1" => "1"
      root_block_device.0.delete_on_termination: "true" => "true"
      root_block_device.0.iops:                  "0" => <computed>
      root_block_device.0.volume_size:           "8" => "8"
      root_block_device.0.volume_type:           "gp2" => "gp2"
      security_groups.#:                         "1" => "1"
      security_groups.476553129:                 "sg-Z" => "sg-Z"
      user_data:                                 "1c89da19319c5107aa54a25cf261bc5ddbfa960c" => "d21405a1c46c871a9eacdef7b8d80b40532b0e23" (forces new resource)


Plan: 2 to add, 0 to change, 2 to destroy.
```

And a trace of this operation encrypted with Hashicorp PGP key: https://gist.github.com/M1ke/9e8a6abac5cafcd0ba152b2853cf25ec

This was run using `--target` to keep it specific, but the same fail occurs with a general run.

**Comment by M1ke** [comment]
I can do that, though the plan does indicate that destroy would happen
before create. With the apply I presume I can run it and then say "no" to
avoid actual downtime?

**Comment by jbardin** [comment]
@M1ke, 

Running `apply` will show the same output, since this is the plan output. The plan itself doesn't indicate that destroy would happen first, are you assuming that from the logs?

**Comment by M1ke** [comment]
Hi @jbardin 

The plan says `-/+ destroy and then create replacement` which is what leads me to believe that destroy would happen first. Have I misunderstood this?

I can confirm that running `terraform apply` gives me the exact same output as in the plan, indicating `-/+` for resources where I'd expect to see `+/-`

**Comment by jbardin** [comment]
Hi @M1ke,

I see the confusion. There is only one diff symbol for resource replacement, which is `-/+` -- it's not meant to imply order, but I now see the line on the top of the diff that could certainly be taken that way.

**Comment by jbardin** [comment]
The output has changed over time to provide more detail and clarity, and will continue to evolve, especially in the next major release. This probably won't be taken care of in 0.11, but I'll leave this open to make sure it's clear in the next version.

**Comment by jbardin** [comment]
Oh to answer your question, I don't think the `-/+` symbol has changed, but the line of text at the top was added for the common case of resource replacement early on in 0.10.

**Comment by apparentlymart** [comment]
Confirming @jbardin's suspicion there, indeed the `-/+` symbol has always been used for both "destroy and then create" and "create and then destroy", but we added the symbol key at the top in order to explain what these symbols mean.

Unfortunately right now the diff renderer does not have enough information to show a different symbol for the `create_before_destroy` case, but I agree that'd be a good change and we can look at what it'd take to make that happen in a future release.

**Comment by apparentlymart** [comment]
Hi again, all!

I just tested this with v0.12.0-alpha2 and found that it is now _closer_ but there is still one small bug to fix:

```
$ terraform apply
null_resource.foo: Refreshing state... [id=7153044459101558325]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
-/+ create replacement and then destroy prior

Terraform will perform the following actions:

  # null_resource.foo must be replaced
+/- resource "null_resource" "foo" {
      ~ id       = "7153044459101558325" -> (known after apply)
      ~ triggers = {
          ~ "foo" = "bar" -> "baz"
        }
    }

Plan: 1 to add, 0 to change, 1 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

null_resource.foo: Creating...
null_resource.foo: Creation complete after 0s [id=5028423547002120744]
null_resource.foo: Destroying... [id=7153044459101558325]
null_resource.foo: Destruction complete after 0s

Apply complete! Resources: 1 added, 1 changed, 0 destroyed.
```

It's showing "create replacement and then destroy prior" as the message in the legend, and using "+/-" as the symbol correctly in the main diff, but the legend is still incorrectly using the "-/+" symbol.

I took a quick peek at the code and I see why it is -- the legend icons are being printed by some older code that doesn't know this new symbol yet -- so we should be able to take care of this before v0.12.0 final.

**Comment by renaldrozario** [comment]
@apparentlymart  create before destroy does it work in openstack? we are running octa

**Comment by apparentlymart** [comment]
Looking at my last comment again with fresh eyes I see that there's another bug hiding there:

```
Apply complete! Resources: 1 added, 1 changed, 0 destroyed.
```

Looks like the destroy was incorrectly counted under "changed" instead of "destroyed", probably again because that older codepath isn't familiar with "create and then destroy" as an action verb and so just counted it as an update instead.

**Comment by radeksimko** [comment]
> Looks like the destroy was incorrectly counted under "changed" instead of "destroyed"

I'm fairly confident this was fixed in #19610

**Comment by radeksimko** [comment]
> It's showing "create replacement and then destroy prior" as the message in the legend, and using "+/-" as the symbol correctly in the main diff, but the legend is still incorrectly using the "-/+" symbol.

Fix pending review in #19642 😉
